# stuff from curriculum

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("untidy-data/treatments.csv")
df

treatmenta  treatmentb  treatmentc
0    John Smith         NaN           2           0
1      Jane Doe        16.0          11           3
2  Mary Johnson         3.0           1           4

In [3]:
df.columns = ['name', 'a', 'b', 'c']
df

name     a   b  c
0    John Smith   NaN   2  0
1      Jane Doe  16.0  11  3
2  Mary Johnson   3.0   1  4

In [4]:
df.melt(id_vars=['name'], var_name='treatment', value_name='response')

name treatment  response
0    John Smith         a       NaN
1      Jane Doe         a      16.0
2  Mary Johnson         a       3.0
3    John Smith         b       2.0
4      Jane Doe         b      11.0
5  Mary Johnson         b       1.0
6    John Smith         c       0.0
7      Jane Doe         c       3.0
8  Mary Johnson         c       4.0

In [5]:
#incorrectly had a melt placed on it
students = pd.read_csv('untidy-data/students.csv')
students.head(3)

date                 var         val
0  2019-02-04   n_late_from_break     4.02812
1  2019-02-04  coffee_consumption  5255.40974
2  2019-02-04      classroom_temp    67.00000

In [6]:
students.pivot(index='date', columns='var')

val                                     
var        classroom_temp coffee_consumption n_late_from_break
date                                                          
2019-02-04           67.0        5255.409740          4.028120
2019-02-05           73.0        8603.704719          2.101998
2019-02-06           81.0        1801.498050          4.941244
2019-02-07           62.0        9282.959741          1.419342
2019-02-08           72.0        7558.270659          1.808919
2019-02-11           60.0        5731.008713          4.831584
2019-02-12           52.0        9547.673484          3.550830
2019-02-13           56.0        7114.931847          2.550362
2019-02-14           72.0        1977.295513          3.507548
2019-02-15           75.0        2254.674679          2.114379
2019-02-18           55.0        4819.981878          2.249609
2019-02-19           82.0        8628.442171          4.646607
2019-02-20           58.0        6905.019187          1.309552
2019-02-21           79.0        1414.705565          2.132542
2019-02-22           68.0        8894.846285          1.745758
2019-02-25           84.0        5982.280540          2.344274
2019-02-26           54.0        9121.391375          4.620919
2019-02-27           88.0        4412.125903          1.860681
2019-02-28           81.0        1693.233409          1.342123

In [7]:
students.set_index(['date', 'var']).unstack(level=1)

val                                     
var        classroom_temp coffee_consumption n_late_from_break
date                                                          
2019-02-04           67.0        5255.409740          4.028120
2019-02-05           73.0        8603.704719          2.101998
2019-02-06           81.0        1801.498050          4.941244
2019-02-07           62.0        9282.959741          1.419342
2019-02-08           72.0        7558.270659          1.808919
2019-02-11           60.0        5731.008713          4.831584
2019-02-12           52.0        9547.673484          3.550830
2019-02-13           56.0        7114.931847          2.550362
2019-02-14           72.0        1977.295513          3.507548
2019-02-15           75.0        2254.674679          2.114379
2019-02-18           55.0        4819.981878          2.249609
2019-02-19           82.0        8628.442171          4.646607
2019-02-20           58.0        6905.019187          1.309552
2019-02-21           79.0        1414.705565          2.132542
2019-02-22           68.0        8894.846285          1.745758
2019-02-25           84.0        5982.280540          2.344274
2019-02-26           54.0        9121.391375          4.620919
2019-02-27           88.0        4412.125903          1.860681
2019-02-28           81.0        1693.233409          1.342123

In [8]:
tidy_students = students.pivot(index='date', columns='var').reset_index()

tidy_students.head(2)

date            val                                     
var             classroom_temp coffee_consumption n_late_from_break
0    2019-02-04           67.0        5255.409740          4.028120
1    2019-02-05           73.0        8603.704719          2.101998

In [9]:
sales = pd.read_csv('untidy-data/sales.csv')
sales

Product  2016 Sales  2016 PPU  2017 Sales  2017 PPU  2018 Sales  2018 PPU
0       A         673         5         231         7         173         9
1       B         259         3         748         5         186         8
2       C         644         3         863         5         632         5
3       D         508         9         356        11         347        14

In [10]:
sales_melt = sales.melt(id_vars='Product', var_name="year_&_measure")
sales_melt.head()

Product year_&_measure  value
0       A     2016 Sales    673
1       B     2016 Sales    259
2       C     2016 Sales    644
3       D     2016 Sales    508
4       A       2016 PPU      5

In [11]:
year_and_measure_df = sales_melt['year_&_measure'].str.split(' ', expand=True)
year_and_measure_df.columns = ['year', 'measure']
year_and_measure_df.head()

year measure
0  2016   Sales
1  2016   Sales
2  2016   Sales
3  2016   Sales
4  2016     PPU

In [12]:
sales2 = pd.concat([sales_melt, year_and_measure_df], axis=1).drop(columns='year_&_measure')
sales2.head()

Product  value  year measure
0       A    673  2016   Sales
1       B    259  2016   Sales
2       C    644  2016   Sales
3       D    508  2016   Sales
4       A      5  2016     PPU

In [13]:
sales_tidy = sales2.pivot_table(index=['Product', 'year'], columns='measure', values='value')
sales_tidy.head()

measure       PPU  Sales
Product year            
A       2016    5    673
        2017    7    231
        2018    9    173
B       2016    3    259
        2017    5    748

In [14]:
sales_tidy.columns.name = ''
sales_tidy.reset_index(inplace=True)
sales_tidy


Product  year  PPU  Sales
0        A  2016    5    673
1        A  2017    7    231
2        A  2018    9    173
3        B  2016    3    259
4        B  2017    5    748
5        B  2018    8    186
6        C  2016    3    644
7        C  2017    5    863
8        C  2018    5    632
9        D  2016    9    508
10       D  2017   11    356
11       D  2018   14    347

# Exercise 1
## Attendance Data

Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

You should end up with something like this:

`name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64`


In [15]:
df = pd.read_csv('untidy-data/attendance.csv')
df.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [16]:
df['name'] = df['Unnamed: 0']
df = df.drop(columns='Unnamed: 0')
df.head()

2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0          P          T          T          H          P          A   
1          A          P          T          T          T          T   
2          A          T          A          A          H          T   
3          P          T          H          P          P          T   

  2018-01-07 2018-01-08   name  
0          T          T  Sally  
1          A          T   Jane  
2          P          T  Billy  
3          P          P   John

In [17]:
df = df.melt(id_vars='name')
df

name    variable value
0   Sally  2018-01-01     P
1    Jane  2018-01-01     A
2   Billy  2018-01-01     A
3    John  2018-01-01     P
4   Sally  2018-01-02     T
5    Jane  2018-01-02     P
6   Billy  2018-01-02     T
7    John  2018-01-02     T
8   Sally  2018-01-03     T
9    Jane  2018-01-03     T
10  Billy  2018-01-03     A
11   John  2018-01-03     H
12  Sally  2018-01-04     H
13   Jane  2018-01-04     T
14  Billy  2018-01-04     A
15   John  2018-01-04     P
16  Sally  2018-01-05     P
17   Jane  2018-01-05     T
18  Billy  2018-01-05     H
19   John  2018-01-05     P
20  Sally  2018-01-06     A
21   Jane  2018-01-06     T
22  Billy  2018-01-06     T
23   John  2018-01-06     T
24  Sally  2018-01-07     T
25   Jane  2018-01-07     A
26  Billy  2018-01-07     P
27   John  2018-01-07     P
28  Sally  2018-01-08     T
29   Jane  2018-01-08     T
30  Billy  2018-01-08     T
31   John  2018-01-08     P

In [18]:
df = df.drop(columns=['variable'])

In [19]:
df.head()

name value
0  Sally     P
1   Jane     A
2  Billy     A
3   John     P
4  Sally     T

In [20]:
attendance_percentage = {
    'P' : 1,
    'T' : .9,
    'H' : 0.5,
    'A' : 0
}

df['attendance_percentage'] = df.value.map(attendance_percentage)
df.head(3)

name value  attendance_percentage
0  Sally     P                    1.0
1   Jane     A                    0.0
2  Billy     A                    0.0

In [21]:
df = df.drop(columns='value')

In [22]:
df.groupby('name')['attendance_percentage'].mean()

name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: attendance_percentage, dtype: float64

# Exercise 2
## Coffee Levels

    a. Read the coffee_levels.csv file.

In [23]:
df = pd.read_csv('untidy-data/coffee_levels.csv')

    b. Transform the data so that each carafe is in it's own column.

In [24]:
df.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [25]:
df['x'] = df.coffee_carafe == 'x'
df['y'] = df.coffee_carafe == 'y'
df['z'] = df.coffee_carafe == 'z'

df = df.drop(columns='coffee_carafe')

In [26]:
df.head()

hour  coffee_amount     x      y      z
0     8       0.816164  True  False  False
1     9       0.451018  True  False  False
2    10       0.843279  True  False  False
3    11       0.335533  True  False  False
4    12       0.898291  True  False  False

    c. Is this the best shape for the data?

In [27]:
df.shape

(30, 5)

I think that it is descent. It could probably be better, because rn i have those bools as their carafe indicators.

# Exercise 3
## Cake Recipes

    a. Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [28]:
df = pd.read_csv('untidy-data/cake_recipes.csv')

    b. Tidy the data as necessary.

In [29]:
df.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [30]:
cut = df["recipe:position"].str.split(':').reset_index()
cut.head()
df['position'] = cut['recipe:position'].str[1]
df['recipe'] = cut['recipe:position'].str[-2]
df = df.drop(columns='recipe:position')
df.head()

225        250        275        300 position recipe
0  61.738655  53.912627  74.414730  98.786784   bottom      a
1  51.709751  52.009735  68.576858  50.228470      top      a
2  57.095320  61.904369  61.196980  99.248541   bottom      b
3  82.455004  95.224151  98.594881  58.169349      top      b
4  96.470207  52.001358  92.893227  65.473084   bottom      c

In [31]:
df_melt = df.melt(id_vars=['recipe', 'position'], var_name='temperature', value_name='rating')
df_melt.head()

recipe position temperature     rating
0      a   bottom         225  61.738655
1      a      top         225  51.709751
2      b   bottom         225  57.095320
3      b      top         225  82.455004
4      c   bottom         225  96.470207

    c. Which recipe, on average, is the best? recipe b

In [32]:
df_melt.groupby('recipe').rating.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: rating, dtype: float64

#### B is the best cake on average

    d. Which oven temperature, on average, produces the best results? 275

In [33]:
df_melt.groupby('temperature').rating.mean()

temperature
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: rating, dtype: float64

#### 275 on avg is producing the best results

    e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [34]:
df_melt.groupby(['recipe', 'position','temperature']).rating.mean().sort_values(ascending=False)

recipe  position  temperature
b       bottom    300            99.248541
a       bottom    300            98.786784
b       top       275            98.594881
d       top       225            96.873178
c       bottom    225            96.470207
b       top       250            95.224151
c       bottom    275            92.893227
        top       275            92.098049
                  250            82.795477
b       top       225            82.455004
d       top       250            76.101363
a       bottom    275            74.414730
c       top       225            71.306308
a       top       275            68.576858
c       bottom    300            65.473084
b       bottom    250            61.904369
a       bottom    225            61.738655
b       bottom    275            61.196980
d       top       275            59.571620
        bottom    250            58.670419
b       top       300            58.169349
        bottom    225            57.095320
d       bottom    300   

#### recipe b, bottom postion and 300 temp cake has thhe highest score